File: 279. Perfect Squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> dp(n+1, INT_MAX);
        for (int i = 1; i*i <= n; i++)
            dp[i*i] = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j*j < i; j++) 
                dp[i] = min(dp[i], dp[i-j*j]+1);
        }
        return dp[n];
    }
};

// dp[i] 表示组成i的最少的完全平方数个数, dp[i] = min(dp[i], )
//动态方程的意思是：对于每个 i ，比 i 小一个完全平方数的那些数中最小的个数+1就是所求，也就是 dp [ i - j * j ] + 1 。

File: go/src/典型问题/DP/279. Perfect Squares/279. Perfect Squares.go
package main

import (
	"fmt"
	"math"
)

func numSquares(n int) int {
	nums := []int{}
	for i := 1; i*i <= n; i++ {
		nums = append(nums, i*i)
	}

	dp := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dp[i] = math.MaxInt32
	}
	for _, num := range nums {
		for i := num; i <= n; i++ {
			dp[i] = min(dp[i], dp[i-num] + 1)
		}
	}
	return dp[n]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(numSquares(12))
	fmt.Println(numSquares(13))
}